src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -135,40 +152,92 @@ public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams reques
                 Diagnostics = mappedDiagnostics,
                 HostDocumentVersion = documentVersion,
             };
+        }
+
+        private async Task<Diagnostic[]> FilterHTMLDiagnosticsAsync(
+            Diagnostic[] unmappedDiagnostics,
+            RazorCodeDocument codeDocument,
+            DocumentSnapshot documentSnapshot)
+        {
+            var sourceText = await documentSnapshot.GetTextAsync();
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var classifiedSpans = syntaxTree.GetClassifiedSpans();
 
-            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
-            static bool CanDiagnosticBeFiltered(RazorLanguageKind kind, Diagnostic d)
+            var processedAttributes = new Dictionary<TextSpan, bool>();
+
+            var filteredDiagnostics = unmappedDiagnostics
+                .Where(d =>
+                    // TODO: undesired HTML Diagnostic codes
+                    // (https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401) ||
+                    !InAttributeContainingCSharp(d, sourceText, syntaxTree))
+                .ToArray();
+
+            return filteredDiagnostics;
+
+            bool InAttributeContainingCSharp(Diagnostic d, SourceText sourceText, RazorSyntaxTree syntaxTree)
             {
-                return kind == RazorLanguageKind.CSharp ?
-                    CanCSharpDiagnosticBeFiltered(d) :
-                    CanHTMLDiagnosticBeFiltered(d);
+                // Examine the _end_ of the diagnostic to see if we're at the
+                // start of an (im/ex)plicit expression. Looking at the start
+                // of the diagnostic isn't sufficient.
+                var absoluteIndex = d.Range.End.GetAbsoluteIndex(sourceText);
+                var change = new SourceChange(absoluteIndex, 0, string.Empty);
+                var owner = syntaxTree.Root.LocateOwner(change);
+
+                while (owner != null)
+                {
+                    if (owner is MarkupAttributeBlockSyntax markupAttributeNode)
```

Comment:
I forget if this also encompasses TagHelper attributes. If not you'll also need to add that to this check

Reply:
So a TagHelper attribute with a HTML diagnostic? If so, do you have an example in mind?

Reply:
TagHelper attributes can overlap HTML attributes to augment tags. For instance, the [ImageTagHelper](https://github.com/dotnet/aspnetcore/blob/1f4a3eb57fa2c2f64d3acde6ca7c04fb0661b282/src/Mvc/Mvc.TagHelpers/src/ImageTagHelper.cs#L83) in non-Blazor scenarios also binds the "src" attribute and does magic with it.

Since TagHelpers are extendable you can technically have that overlap on any attribute which means that the HTML side of the world will treat those attributes as normal HTML attributes and log errors on them accordingly. If you want an easy way to repro you can write a TagHelper that binds the style attribute. Aka:

```C#
// Will apply to the <div> tag
public class DivTagHelper : TagHelper
{
    // Will bind the "style" attribute making it a TagHelper attribute
    public string Style { get; set; }
}
```

```
@addTagHelper *, MyAppAssembly
```

Reply:
Thanks for the clarification and call out. Updated to ensure taghelper attributes are also covered:


![image](https://user-images.githubusercontent.com/14852843/105393976-012b7700-5beb-11eb-84d0-57a5386dc859.png)